assembly/src/lib.rs

```diff
@@ -65,28 +87,37 @@ impl Assembler {
     // CONSTRUCTORS
     // --------------------------------------------------------------------------------------------
     /// Returns a new instance of [Assembler] instantiated with empty module map.
-    ///
-    /// Debug related decorators are added to span blocks when debug mode is enabled.
-    pub fn new(in_debug_mode: bool) -> Self {
+    pub fn new() -> Self {
         Self {
-            stdlib: StdLibrary::default(),
+            module_provider: Box::new(()),
             parsed_modules: BTreeMap::default(),
             kernel_procs: BTreeMap::default(),
             kernel: Kernel::default(),
-            in_debug_mode,
+            in_debug_mode: false,
         }
     }
```

Comment:
I like this new constructor approach! A couple of suggestions - but they are pretty minor - so, feel free to keep it as is:

- I wonder if we should get rid of `with_debug_mode()` and keep the signature of `new()` as it was before (i.e., `fn new(in_debug_mode: bool) -> Self`. We'd still keep `with_kernel()` and `with_module_provider()` as you have now.
- very minor nit: I'd move `with_kernel()` after `with_module_provider()`.

Reply:
Regarding the order, I always tend to keep alphabetical order so its easier to seek, but in this case, logically, it makes more sense to have the module provider before the kernel. Done!

About the second point, we have the [builder pattern docs](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html) that says we should use the `new` only with the minimal viable arguments to create the type. However, since this is an hybrid (i.e. we don't have a distinct type for the builder), then we don't necessarily have to follow these rules. wdyt?

Reply:
Regarding the second point: I don't have a strong preference - so, let's keep it as you have now.